File: 0152.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        
        int maxProd = nums[0];
        int curMax = nums[0];
        int curMin = nums[0];
        
        for(int i=1; i<nums.size(); i++) {
            int temp = curMax;
            int tempCurMax = max(nums[i], curMax * nums[i]);
            curMax = max(tempCurMax, curMin * nums[i]);
            
            int tempCurMin = min(nums[i], temp * nums[i]);
            curMin = min(tempCurMin, curMin * nums[i]);
            
            if(curMax > maxProd) {
                maxProd = curMax;
            }
            
        }
        
        return maxProd;
    }
};

File: 0152.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        maxProd = nums[0];
        curMax = nums[0];
        curMin = nums[0];
        
        for i in range (1, len(nums)):
            temp = curMax
            curMax = max(nums[i], curMax * nums[i], curMin * nums[i])
            curMin = min(nums[i], temp * nums[i], curMin * nums[i])
            
            if curMax > maxProd:
                maxProd = curMax
            
        return maxProd;
        